File: agencia-frontend/src/app/services/servicio.service.ts
import { Injectable } from '@angular/core';
import { Apollo, gql } from 'apollo-angular';
import { Observable, map } from 'rxjs';
import { Servicio, CreateServicioInput, UpdateServicioInput } from '../models/servicio.model';

const GET_ALL_SERVICIOS = gql`
  query GetAllServicios {
    getAllServicios {
      id
      proveedorId
      proveedor {
        id
        nombreEmpresa
        tipoServicio
      }
      tipoServicio
      nombreServicio
      descripcion
      destinoCiudad
      destinoPais
      precioCosto
      precioVenta
      isAvailable
    }
  }
`;

const GET_SERVICIO_BY_ID = gql`
  query GetServicioById($id: ID!) {
    getServicioById(id: $id) {
      id
      proveedorId
      proveedor {
        id
        nombreEmpresa
        tipoServicio
      }
      tipoServicio
      nombreServicio
      descripcion
      destinoCiudad
      destinoPais
      precioCosto
      precioVenta
      isAvailable
    }
  }
`;

const GET_SERVICIOS_BY_PROVEEDOR = gql`
  query GetServiciosByProveedorId($proveedorId: ID!) {
    getServiciosByProveedorId(proveedorId: $proveedorId) {
      id
      proveedorId
      tipoServicio
      nombreServicio
      descripcion
      destinoCiudad
      destinoPais
      precioCosto
      precioVenta
      isAvailable
    }
  }
`;

const GET_SERVICIOS_BY_TIPO = gql`
  query GetServiciosByTipo($tipoServicio: String!) {
    getServiciosByTipo(tipoServicio: $tipoServicio) {
      id
      proveedorId
      proveedor {
        id
        nombreEmpresa
      }
      tipoServicio
      nombreServicio
      destinoCiudad
      destinoPais
      precioVenta
      isAvailable
    }
  }
`;

const SEARCH_SERVICIOS = gql`
  query SearchServicios($searchTerm: String!) {
    searchServicios(searchTerm: $searchTerm) {
      id
      proveedorId
      proveedor {
        id
        nombreEmpresa
      }
      tipoServicio
      nombreServicio
      descripcion
      destinoCiudad
      destinoPais
      precioVenta
      isAvailable
    }
  }
`;

const CREATE_SERVICIO = gql`
  mutation CreateServicio($input: CreateServicioInput!) {
    createServicio(input: $input) {
      id
      proveedorId
      tipoServicio
      nombreServicio
      descripcion
      destinoCiudad
      destinoPais
      precioCosto
      precioVenta
      isAvailable
    }
  }
`;

const UPDATE_SERVICIO = gql`
  mutation UpdateServicio($id: ID!, $input: UpdateServicioInput!) {
    updateServicio(id: $id, input: $input) {
      id
      proveedorId
      tipoServicio
      nombreServicio
      descripcion
      destinoCiudad
      destinoPais
      precioCosto
      precioVenta
      isAvailable
    }
  }
`;

const DELETE_SERVICIO = gql`
  mutation DeleteServicio($id: ID!) {
    deleteServicio(id: $id)
  }
`;

@Injectable({
  providedIn: 'root'
})
export class ServicioService {

  constructor(private apollo: Apollo) {}

  getAllServicios(): Observable<Servicio[]> {
    return this.apollo.query<{ getAllServicios: Servicio[] }>({
      query: GET_ALL_SERVICIOS,
      fetchPolicy: 'network-only'
    }).pipe(
      map(result => result.data?.getAllServicios || [])
    );
  }

  getServicioById(id: string): Observable<Servicio> {
    return this.apollo.query<{ getServicioById: Servicio }>({
      query: GET_SERVICIO_BY_ID,
      variables: { id },
      fetchPolicy: 'network-only'
    }).pipe(
      map(result => result.data!.getServicioById)
    );
  }

  getServiciosByProveedorId(proveedorId: string): Observable<Servicio[]> {
    return this.apollo.query<{ getServiciosByProveedorId: Servicio[] }>({
      query: GET_SERVICIOS_BY_PROVEEDOR,
      variables: { proveedorId },
      fetchPolicy: 'network-only'
    }).pipe(
      map(result => result.data?.getServiciosByProveedorId || [])
    );
  }

  getServiciosByTipo(tipoServicio: string): Observable<Servicio[]> {
    return this.apollo.query<{ getServiciosByTipo: Servicio[] }>({
      query: GET_SERVICIOS_BY_TIPO,
      variables: { tipoServicio },
      fetchPolicy: 'network-only'
    }).pipe(
      map(result => result.data?.getServiciosByTipo || [])
    );
  }

  searchServicios(searchTerm: string): Observable<Servicio[]> {
    return this.apollo.query<{ searchServicios: Servicio[] }>({
      query: SEARCH_SERVICIOS,
      variables: { searchTerm },
      fetchPolicy: 'network-only'
    }).pipe(
      map(result => result.data?.searchServicios || [])
    );
  }

  createServicio(input: CreateServicioInput): Observable<Servicio> {
    return this.apollo.mutate<{ createServicio: Servicio }>({
      mutation: CREATE_SERVICIO,
      variables: { input },
      refetchQueries: [{ query: GET_ALL_SERVICIOS }]
    }).pipe(
      map(result => result.data!.createServicio)
    );
  }

  updateServicio(id: string, input: UpdateServicioInput): Observable<Servicio> {
    return this.apollo.mutate<{ updateServicio: Servicio }>({
      mutation: UPDATE_SERVICIO,
      variables: { id, input },
      refetchQueries: [{ query: GET_ALL_SERVICIOS }]
    }).pipe(
      map(result => result.data!.updateServicio)
    );
  }

  deleteServicio(id: string): Observable<boolean> {
    return this.apollo.mutate<{ deleteServicio: boolean }>({
      mutation: DELETE_SERVICIO,
      variables: { id },
      refetchQueries: [{ query: GET_ALL_SERVICIOS }]
    }).pipe(
      map(result => result.data!.deleteServicio)
    );
  }
}

File: agencia-backend/src/main/java/com/agencia/agencia_backend/service/ServicioService.java
package com.agencia.agencia_backend.service;

import com.agencia.agencia_backend.dto.CreateServicioInput;
import com.agencia.agencia_backend.dto.UpdateServicioInput;
import com.agencia.agencia_backend.model.Servicio;
import com.agencia.agencia_backend.repository.ServicioRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class ServicioService {

    @Autowired
    private ServicioRepository servicioRepository;

    /**
     * Obtener todos los servicios
     */
    public List<Servicio> getAllServicios() {
        return servicioRepository.findAll();
    }

    /**
     * Obtener servicio por ID
     */
    public Optional<Servicio> getServicioById(String id) {
        return servicioRepository.findById(id);
    }

    /**
     * Obtener servicios por proveedor ID
     */
    public List<Servicio> getServiciosByProveedorId(String proveedorId) {
        return servicioRepository.findByProveedorId(proveedorId);
    }

    /**
     * Obtener servicios por tipo
     */
    public List<Servicio> getServiciosByTipo(String tipoServicio) {
        return servicioRepository.findByTipoServicio(tipoServicio);
    }

    /**
     * Obtener servicios por destino
     */
    public List<Servicio> getServiciosByDestino(String destinoCiudad) {
        return servicioRepository.findByDestinoCiudad(destinoCiudad);
    }

    /**
     * Buscar servicios por nombre
     */
    public List<Servicio> searchServicios(String searchTerm) {
        List<Servicio> servicios = servicioRepository.findAll();
        
        return servicios.stream()
            .filter(servicio -> {
                String nombreServicio = servicio.getNombreServicio().toLowerCase();
                String descripcion = servicio.getDescripcion() != null ? 
                    servicio.getDescripcion().toLowerCase() : "";
                String destino = servicio.getDestinoCiudad() != null ? 
                    servicio.getDestinoCiudad().toLowerCase() : "";
                return nombreServicio.contains(searchTerm.toLowerCase()) || 
                       descripcion.contains(searchTerm.toLowerCase()) ||
                       destino.contains(searchTerm.toLowerCase());
            })
            .toList();
    }

    /**
     * Crear nuevo servicio
     */
    public Servicio createServicio(CreateServicioInput input) {
        Servicio servicio = new Servicio();
        servicio.setProveedorId(input.getProveedorId());
        servicio.setTipoServicio(input.getTipoServicio());
        servicio.setNombreServicio(input.getNombreServicio());
        servicio.setDescripcion(input.getDescripcion());
        servicio.setDestinoCiudad(input.getDestinoCiudad());
        servicio.setDestinoPais(input.getDestinoPais());
        servicio.setPrecioCosto(input.getPrecioCosto());
        servicio.setPrecioVenta(input.getPrecioVenta());
        servicio.setIsAvailable(input.getIsAvailable());

        return servicioRepository.save(servicio);
    }

    /**
     * Actualizar servicio existente
     */
    public Servicio updateServicio(String id, UpdateServicioInput input) {
        Servicio servicio = servicioRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Servicio no encontrado con ID: " + id));

        // Actualizar solo los campos que no son null
        if (input.getProveedorId() != null) {
            servicio.setProveedorId(input.getProveedorId());
        }
        if (input.getTipoServicio() != null) {
            servicio.setTipoServicio(input.getTipoServicio());
        }
        if (input.getNombreServicio() != null) {
            servicio.setNombreServicio(input.getNombreServicio());
        }
        if (input.getDescripcion() != null) {
            servicio.setDescripcion(input.getDescripcion());
        }
        if (input.getDestinoCiudad() != null) {
            servicio.setDestinoCiudad(input.getDestinoCiudad());
        }
        if (input.getDestinoPais() != null) {
            servicio.setDestinoPais(input.getDestinoPais());
        }
        if (input.getPrecioCosto() != null) {
            servicio.setPrecioCosto(input.getPrecioCosto());
        }
        if (input.getPrecioVenta() != null) {
            servicio.setPrecioVenta(input.getPrecioVenta());
        }
        if (input.getIsAvailable() != null) {
            servicio.setIsAvailable(input.getIsAvailable());
        }

        return servicioRepository.save(servicio);
    }

    /**
     * Eliminar servicio (hard delete)
     */
    public boolean deleteServicio(String id) {
        Servicio servicio = servicioRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Servicio no encontrado con ID: " + id));

        servicioRepository.delete(servicio);
        return true;
    }
}
